sw/device/lib/dif/dif_keymgr.c

```diff
@@ -5,3 +5,723 @@
+  *status_codes = bitfield_bit32_write(0, 0, is_idle);
+
+  if (has_error) {
+    // Read and clear ERR_CODE register (rw1c).
+    uint32_t reg_err_code = mmio_region_read32(keymgr->params.base_addr,
+                                               KEYMGR_ERR_CODE_REG_OFFSET);
+    mmio_region_write32(keymgr->params.base_addr, KEYMGR_ERR_CODE_REG_OFFSET,
+                        reg_err_code);
+    // Error bits start from bit 1 in `dif_keymgr_status_codes_t`.
+    // Note: The mask is hardcoded since it is not auto generated yet.
+    const bitfield_field32_t kErrorBitfield = (bitfield_field32_t){
+        .mask = 0xF,
+        .index = 1,
+    };
+    if (reg_err_code > kErrorBitfield.mask || reg_err_code == 0) {
+      return kDifKeymgrError;
+    }
+    *status_codes =
+        bitfield_field32_write(*status_codes, kErrorBitfield, reg_err_code);
+  }
+
+  return kDifKeymgrOk;
+}
+
+dif_keymgr_result_t dif_keymgr_get_state(const dif_keymgr_t *keymgr,
+                                         dif_keymgr_state_t *state) {
+  if (keymgr == NULL || state == NULL) {
+    return kDifKeymgrBadArg;
+  }
+
+  uint32_t reg_state = mmio_region_read32(keymgr->params.base_addr,
+                                          KEYMGR_WORKING_STATE_REG_OFFSET);
+
+  switch (bitfield_field32_read(reg_state, KEYMGR_WORKING_STATE_STATE_FIELD)) {
+    case KEYMGR_WORKING_STATE_STATE_VALUE_RESET:
+      *state = kDifKeymgrStateReset;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_INIT:
+      *state = kDifKeymgrStateInitialized;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_CREATOR_ROOT_KEY:
+      *state = kDifKeymgrStateCreatorRootKey;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_OWNER_INTERMEDIATE_KEY:
+      *state = kDifKeymgrStateOwnerIntermediateKey;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_OWNER_KEY:
+      *state = kDifKeymgrStateOwnerRootKey;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_DISABLED:
+      *state = kDifKeymgrStateDisabled;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_INVALID:
+      *state = kDifKeymgrStateInvalid;
+      break;
+    default:
+      return kDifKeymgrError;
+  }
+
+  return kDifKeymgrOk;
+}
+
+dif_keymgr_lockable_result_t dif_keymgr_generate_identity_seed(
+    const dif_keymgr_t *keymgr) {
+  if (keymgr == NULL) {
+    return kDifKeymgrLockableBadArg;
+  }
+
+  if (!is_idle(keymgr)) {
+    return kDifKeymgrLockableLocked;
+  }
+
+  start_operation(keymgr, (start_operation_params_t){
+                              .dest = KEYMGR_CONTROL_DEST_SEL_VALUE_NONE,
+                              .op = KEYMGR_CONTROL_OPERATION_VALUE_GENERATE_ID,
+                          });
+
+  return kDifKeymgrLockableOk;
+}
+
+dif_keymgr_lockable_result_t dif_keymgr_generate_versioned_key(
+    const dif_keymgr_t *keymgr, dif_keymgr_versioned_key_params_t params) {
+  if (keymgr == NULL) {
+    return kDifKeymgrLockableBadArg;
+  }
+
+  start_operation_params_t hw_op_params;
```

Comment:
It might be worth initializing this (e.g. `= {0}`) to something just in case someone adds more fields in the future.

Reply:
Maybe we should just use designated initializers below instead? 🤔